refactor logs to not use internal data structures

MHBauer opened the pull request:
In preparation for future work to pull the APIs into a separate deliverable, we need to not make references back into the daemon or other core engine code. The logs command referenced the daemon-struct for a container. This work attempts to move the code that does that into the daemon itself.

Is there any worry about not bailing as early on failure? 

Given the use of daemon.Get() for checking if a container exists, is the interest in a new function to do that and return some basic information? Looking at some of this stuff it feels like something missing, but I'm not sure it would have much use outside of the couple locations that already exist. 

see also #16361 

cc @icecrime @calavera 
& @cpuguy83 because he seemed interested in the last one.

duglin commented:
I think moving the check into the daemon code is fine. 

Not sure of the need for a new func since Get() will return the container and people should be able to just ask it for any info they need. But perhaps I'm not understanding your question.

cpuguy83 commented:
The thing to watch out for is `outStream := ioutils.NewWriteFlusher(w)`
Once that hits, we can't give a proper response back for non-OK responses, ie if the container is not found.

cpuguy83 commented:
Did you find any tests related to logs API response codes?

duglin commented:
@cpuguy83 I thought about that as well. I tested this PR to make sure a 404 is returned when we can't find the container (both streaming and not-streaming cases), so it looked good.  @MHBauer I agree with @cpuguy83, let's make sure we have testcases (api tests) that check for 404 to make sure we don't break this in the future.

MHBauer commented:
@cpuguy83 @duglin Do these tests look alright? I'm checking for the non-existent container case. Are there other cases you would like to see?

MHBauer commented:
fixed the cli test which was a bad test.

I'm still stuck on how to resolve how to send the 404 and not a 200.

cpuguy83 commented on integration-cli/docker_cli_logs_test.go:
```diff
@@ -374,3 +374,11 @@ func (s *DockerSuite) TestLogsFollowGoroutinesNoOutput(c *check.C) {
 }
+
+func (s *DockerSuite) TestLogsCLIContainerNotFound(c *check.C) {
+	testRequires(c, DaemonIsLinux)
```
I don't think this is required for this test

cpuguy83 commented on integration-cli/docker_api_logs_test.go:
```diff
@@ -82,3 +82,11 @@ func (s *DockerSuite) TestLogsApiFollowEmptyOutput(c *check.C) {
 }
+
+func (s *DockerSuite) TestLogsAPIContainerNotFound(c *check.C) {
+	testRequires(c, DaemonIsLinux)
```
I don't think this is required for this test

cpuguy83 commented:
So generally I think it's worse to have HTTP concerns in the daemon than the other way around.
Seems the problem is really around how we handle the write flusher.

duglin commented:
@cpuguy83 do you know why we need to send the nil?  I'm trying to understand why we would need to send the http header right away in the happy case.  I would think the CLI would just wait around for it to send some data.

MHBauer commented:
I agree that http stuff shouldn't be in the daemon. I can't figure out how to get the 'Write(nil)' out.

LK4D4 commented on daemon/logs.go:
```diff
@@ -30,13 +30,24 @@ type ContainerLogsConfig struct {
 // ContainerLogs hooks up a container's stdout and stderr streams
 // configured with the given struct.
-func (daemon *Daemon) ContainerLogs(container *Container, config *ContainerLogsConfig) error {
+func (daemon *Daemon) ContainerLogs(prefixOrName string, config *ContainerLogsConfig) error {
```
this var name is too descriptive. name will be enough.

duglin replied:
LOL that's a great/funny comment.

MHBauer commented:
I rearranged the code so that we don't set up the stdcopy embedded streams until we are done with error condition checking.

@duglin floated the idea of inverting the call somewhat, by making it return a callback. That way the impl can do some basic checking before returning a function and error, which the api can check, before continuing the standard processing. I think the code rearranging helps that. 

I disagree with the test in docker_cli_daemon_test.go. If you run the logs command, and don't get logs, that's an error, even if it is because you are using the none-driver.

The fprintf doesn't add anything more than the normal http handler, so that stayed removed.

cpuguy83 commented on api/common.go:
```diff
@@ -17,7 +17,7 @@ import (
 const (
 	// Version of Current REST API
-	Version version.Version = "1.21"
+	Version version.Version = "1.20"
```
How come this was changed?

MHBauer replied:
my bad. putting it back. Needed to test with a newer server and the client doesn't like that.

LK4D4 commented on daemon/logs.go:
```diff
@@ -58,7 +55,21 @@ func (daemon *Daemon) ContainerLogs(container *Container, config *ContainerLogsC
+		errStream = stdcopy.NewStdWriter(outStream, stdcopy.Stderr)
+		outStream = stdcopy.NewStdWriter(outStream, stdcopy.Stdout)
+	}
+	config.OutStream = outStream
```
btw I don't see where `config.OutStream` value used below. Maybe we don't need to modify config here?

MHBauer replied:
It's a dead write if I take out the fprintf in api/server/container.go

Yea, since I'm looking at the attach code, it looks like this block was in the past copied and pasted from there, where thy do write it back before reading it out again. I'll change it here.

duglin commented:
did you want to refactor this one more (to use daemon.Exists() ) or is it ready?

MHBauer commented:
I did not think of that. It passes the tests as it is. Do you want it changed to use Exists?

MHBauer commented:
Here's it using exists on top of the old one so people can compare. I would squash before the merge.

MHBauer committed:
refactor logs to not use internal data structures

 - refactor to make it easier to split the api in the future
 - additional tests for non existent container case

Signed-off-by: Morgan Bauer <mbauer@us.ibm.com>

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "logs-refactor" git@github.com:MHBauer/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

duglin commented:
LGTM

ping @jfrazelle @vdemeester

jfrazelle commented:
LGTM

vdemeester commented:
ah, had a small nit on the test, but no biggies :cat:, I can't compete with jess :stuck_out_tongue_closed_eyes:

MHBauer commented:
@vdemeester what is the nit before you forget? I would like to know.

jfrazelle commented:
sorrryyyyy

On Thu, Oct 8, 2015 at 2:11 PM, Vincent Demeester <notifications@github.com>
wrote:

> ah, had a small nit on the test, but no biggies [image: :cat:], I can't
> compete with jess [image: :stuck_out_tongue_closed_eyes:]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/16366#issuecomment-146685939>.
>

jfrazelle commented:
:(

On Thu, Oct 8, 2015 at 2:14 PM, Jessie Frazelle <jess@docker.com> wrote:

> sorrryyyyy
>
> On Thu, Oct 8, 2015 at 2:11 PM, Vincent Demeester <
> notifications@github.com> wrote:
>
>> ah, had a small nit on the test, but no biggies [image: :cat:], I can't
>> compete with jess [image: :stuck_out_tongue_closed_eyes:]
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/docker/docker/pull/16366#issuecomment-146685939>.
>>
>
>
